corpus_analysis/document.py

```diff
@@ -637,3 +640,83 @@ def update_metadata(self, new_metadata):
                         f" not '{new_metadata['date']}'"
                     ) from err
             setattr(self, key, new_metadata[key])
+
+    def get_char_list(self, cutoff_num=10):
+        """
+        given a document object, find a list of characters with their frequency in the novels
+        param: cutoff_num defaults to 10 for the thredshold for cutoffs based on named frequency
+        return: a list of tuples with character names in descending sorted order that occurs
+        more than the cutoff_num times in the document
+        >>> from corpus_analysis import document
+        >>> from pathlib import Path
+        >>> from gender_analysis import common
+        >>> document_metadata = {'author': 'Austen, Jane', 'title': 'Persuasion', 'date': '1818',
+        ...                      'filename': 'austen_persuasion.txt',
+        ...                      'filepath': Path(common.TEST_DATA_PATH, 'sample_novels',
+        ...                                       'texts', 'austen_persuasion.txt')}
+        >>> persuasion = document.Document(document_metadata)
+        >>> persuasion_chars = persuasion.get_char_list(20)
+        >>> persuasion_chars
+        [('Anne', 425), ('Captain Wentworth', 119), ('Lady Russell', 116), ('Charles', 115), ('Mary', 113), ('Sir Walter', 95), ('Elizabeth', 82), ('Elliot', 76), ('Louisa', 66), ('Henrietta', 65), ('Mrs Musgrove', 40), ('Mrs Smith', 36), ('Mrs Clay', 33), ('Miss Elliot', 32), ('Captain Benwick', 32), ('Wentworth', 31), ('Charles Hayter', 29), ('Mrs Croft', 27), ('Benwick', 26), ('Musgrove', 24), ('Uppercross', 23), ('Lady Dalrymple', 22), ('Captain Harville', 21)]
+        >>> len(persuasion_chars)
+        23
+        """
+
+        labels_char = []
+        labels = 'FACILITY,GPE,GSP,LOCATION,ORGANIZATION,PERSON'
+        document = self._load_document_text()
+        sentences = nltk.sent_tokenize(document)
+        for sent in sentences:
+            for chunk in nltk.ne_chunk(nltk.pos_tag(nltk.word_tokenize(sent))):
+                if hasattr(chunk, 'label'):
+                    labels_char.append((chunk.label(), ' '.join(c[0] for c in chunk)))
```

Comment:
Four considerations on these last few lines:
1. What's the purpose of calling `sent_tokenize`? Could we simply call `nltk.word_tokenize(document.text)` rather than iterating through the sentences and passing each sentence into `word_tokenize`?
2. It looks like we're doing some duck typing here in line 671 in that we're checking the `chunk` for the existence of the attribute `label`. I'd prefer to see an explicit type check there (`if isinstance(chunk, nltk.tree.Tree)`) because I think that's more useful to us in the future. As currently written, if you want to find out what `label()` as a method is and does, you'll need to open up a terminal and manually inspect the type of `chunk` (or, of course, just know it based on your familiarity with `nltk`) to find out the class that `label()` is a method on, then go find the relevant docs (in this case, [here](https://www.nltk.org/_modules/nltk/tree.html#Tree)), whereas if we explicitly check the type it saves the developer a step.
3. In general, as a matter of style, I'd highly recommend writing out full variable names. That is, `sentence` instead of `sent` and `label_characters` instead of `label_char`. Each individual instance of this kind of abbreviation isn't a big deal, but when they start to pile up (as in line 672 here and 676 below), it becomes challenging to read through the code change quickly. I think that writing your code such that a subsequent developer is able to skim the changes quickly is important (hence all my comments here).
4. I'd like to see line 670 expanded. That is, I'd like to see each of those `nltk` method calls pulled out into a variable declaration. Something like:
```
for sentence in sentences:
  sentence_tokens = nltk.word_tokenize(sentence)
  sentence_pos_tags = nltk.pos_tag(sentence_tokens)
  sentence_chunks = nltk.ne_chunk(sentence_pos_tags)
  for chunk in sentence_chunks:
    etc.
```
This greatly aids in the subsequent developer's ability to skim the code and understand what's happening.